File: RansomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        m = {}
        for char in magazine:
            if char in m:
                m[char] += 1
            else:
                m[char] = 1
        
        for char in ransomNote:
            if char in m and m[char] > 0:
                m[char] -= 1
            else:
                return False

        return True

    # Time: O(n + m)
    # Space: O(m) m for the number of char in magazine

File: RansomNote.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> map = new HashMap<Character, Integer>();

        for (char c : magazine.toCharArray()){
            if (map.containsKey(c)){
                map.put(c, map.get(c) + 1);
            }else{
                map.put(c, 1);
            }
        }

        for (char c : ransomNote.toCharArray()){
            if (map.containsKey(c) && map.get(c) > 0){
                map.put(c, map.get(c) - 1);
            }else{
                return false;
            }
        }

        return true;
    }
}

// Time: O(n + m)
// Space: O(m) m is the number of char in magazine
